DatacenterIntegration/Portal/PortalUserDataGdprUtilities.psm1

```diff
@@ -0,0 +1,310 @@
+            $curResult = New-Object PSObject
+            $curResult | add-member Noteproperty DirectoryTenantId       $dirId
+            $curResult | add-member Noteproperty UserName                $UserPrincipalName
+            $curResult | add-member Noteproperty ErrorMessage            "No user under tenant directory"
+            $clearUserDataResult += @( $curResult )
+            continue
+        }
+        elseif (([string[]]$userObjectId).Length -gt 1)
+        {
+            Write-Warning "There is one more users retrieved with '$UserPrincipalName' under directory tenant Id $dirId."
+            $curResult = New-Object PSObject
+            $curResult | add-member Noteproperty DirectoryTenantId       $dirId
+            $curResult | add-member Noteproperty UserName                $UserPrincipalName
+            $curResult | add-member Noteproperty ErrorMessage            "User principal name incorrect. One more user accounts under tenant directory"
+            $clearUserDataResult += @( $curResult )
+            continue
+        }
+        else
+        {
+            $params = @{
+                AzsEnvironment      = $adminArmEnv
+                UserObjectId        = $userObjectId
+                DirectoryTenantId   = $dirId
+                AdminSubscriptionId = $adminSubscriptionId
+                AzsAdminArmEndpoint = $AzsAdminArmEndpoint
+            }
+            $curResult = Clear-AzsUserData @params
+            $clearUserDataResult += @( $curResult )
+        }
+    }
+
+    return $clearUserDataResult
+}
+
+<#
+.Synopsis
+   Clear the portal user data under ADFS environment
+#>
+function Clear-AzsUserDataUnderADFS
+{
+    param
+    (
+        # The user credential with which to acquire an access token targeting Graph.
+        [Parameter(Mandatory=$true)]
+        [ValidateNotNull()]
+        [pscredential] $AdfsGraphCred = $null,
+
+        # The Azure Stack administrator credential.
+        [Parameter(Mandatory=$true)]
+        [ValidateNotNull()]
+        [pscredential] $AzsAdminCred = $null,
+
+        # The directory tenant identifier.
+        [Parameter(Mandatory=$true)]
+        [ValidateNotNullOrEmpty()]
+        [string] $DirectoryTenantId,
+
+        # The Azure Stack ARM endpoint URI.
+        [Parameter(Mandatory=$true)]
+        [ValidateNotNullOrEmpty()]
+        [Uri] $AzsAdminArmEndpoint,
+
+        # The user principal name of the account who's user data should be cleared.
+        [Parameter(Mandatory=$true)]
+        [ValidateNotNullOrEmpty()]
+        [string] $UserPrincipalName
+    )
+    #requires -Version 4.0
+    #requires -Module "AzureRM.Profile"
+    #requires -RunAsAdministrator
+
+    Import-Module $PSScriptRoot\..\..\Identity\GraphAPI\GraphAPI.psm1 -Force
+
+    Write-Verbose "Login to Azure Stack Admin ARM..." -Verbose
+    $AzsAdminEnvironmentName = "AzureStackAdmin"
+    $adminArmEnv = Add-AzureRmEnvironment -Name ($AzsAdminEnvironmentName) -ARMEndpoint $AzsAdminArmEndpoint
```

Comment:
@keystroke , I'd like to understand what kind of issue you might be referring to here.

Reply:
I don't think the cmdlet is idempotent; in some code, we first call GET to see if environment already exists, and then either proceed, or remove it to re-add it. In Multitenancy code, we call "add" with -ErrorAction Ignore, and then call GET with -ErrorAction stop to facilitate an "idempotent" initialize, as I think the Add cmdlet will return an error if the env already exists. Haven't checked the latest behavior though, just been using this pattern since the beginning.

Reply:
use the GUID as the azure environment name? @keystroke

Reply:
As talked to @keystroke Add-AzureRmEnvironment is idempotent.